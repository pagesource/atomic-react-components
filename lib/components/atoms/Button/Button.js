// @flow
import React from 'react';
import type { Node } from 'react';
import styled from 'styled-components';
import classnames from 'classnames';

import { getLabelFromChildren } from '../../../core/utils/elementLocator';
import styles from './Button.style';

type Props = {
  className?: string,
  children: Node,
  inheritedStyles?: string,
  type?: string,
  primary?: boolean,
  secondary?: boolean,
  tertiary?: boolean,
  disabled?: boolean,
  ariaLabel?: string,
  elementLocator?: string,
  icon?: Boolean | String,
  iconImage?: String,
};

const Button = ({
  className,
  children,
  inheritedStyles,
  type,
  primary,
  secondary,
  tertiary,
  disabled,
  ariaLabel,
  elementLocator,
  icon,
  iconImage,
  ...others
}: Props): Node => (
  <button
    aria-label={ariaLabel || null}
    className={classnames(className, { 'icon icon-image': iconImage })}
    disabled={disabled}
    type={type}
    data-locator={elementLocator || `button-${type}-${getLabelFromChildren(children)}`}
    {...others}
  >
    {children}
  </button>
);

Button.defaultProps = {
  inheritedStyles: '',
  type: 'button',
  primary: false,
  secondary: false,
  tertiary: false,
  disabled: false,
  ariaLabel: '',
  className: '',
  elementLocator: '',
  icon: false,
  iconImage: '',
};

export default styled(Button)`
  ${styles};
`;
export { Button as ButtonVanilla };
