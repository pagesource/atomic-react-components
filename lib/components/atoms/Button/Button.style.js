import { css } from 'styled-components';

const ButtonStyles = css`
  border: none;
  border-radius: 0;
  position: relative;
  min-height: 48px;
  cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};
  ${props => `
      ${props.theme.colors.CTA_PRIMARY};
      ${props.theme.typography.FONT_CTA};
      padding: 0 ${props.theme.spacing.UNIT_2};
  `};
  ${props => props.secondary && `${props.theme.colors.CTA_SECONDARY}`};
  ${props =>
    props.tertiary
      ? `
    ${props.theme.colors.CTA_TERTIARY};
    ${props.theme.border.CTA_TERTIARY};
  `
      : ``};
  ${props => props.fullWidth && ` width: 100% ;`};

  &:focus:not(.focus-visible) {
    outline: none;
  }

  &.icon {
    border: none;
    background-color: transparent;
    text-indent: -1000px;
    overflow: hidden;
    background-repeat: no-repeat;
    background-position: center;
    background-size: cover;
    min-width: 20px;
    min-height: 20px;
    padding: 0;

    &.icon-image {
      background-image: url('${props => props.iconImage}');
    }
  }

  ${props => (props.inheritedStyles ? props.inheritedStyles : '')};
`;

export default ButtonStyles;
