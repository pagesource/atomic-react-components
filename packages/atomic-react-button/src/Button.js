import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { getLabelFromChildren } from '@xt-pagesource/atomic-react-utils';
import styles from './Button.style';

const Button = ({
  className,
  children,
  inheritedStyles,
  type,
  primary,
  secondary,
  tertiary,
  variation,
  disabled,
  transparent,
  ariaLabel,
  elementLocator,
  ...others
}) => (
  <button
    {...others}
    aria-label={ariaLabel || null}
    className={className}
    disabled={disabled}
    type={type}
    data-locator={
      elementLocator || (type ? `button-${type}-${getLabelFromChildren(children)}` : '')
    }
    variation={variation}
  >
    {children}
  </button>
);

Button.defaultProps = {
  inheritedStyles: '',
  type: 'button',
  primary: false,
  secondary: false,
  tertiary: false,
  disabled: false,
  variation: '',
  transparent: false,
  ariaLabel: '',
  className: '',
  elementLocator: '',
};

Button.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  inheritedStyles: PropTypes.string,
  type: PropTypes.string,
  primary: PropTypes.bool,
  secondary: PropTypes.bool,
  tertiary: PropTypes.bool,
  disabled: PropTypes.bool,
  variation: PropTypes.string,
  transparent: PropTypes.bool,
  ariaLabel: PropTypes.string,
  elementLocator: PropTypes.string,
};

const StyledButton = styled(Button)`
  ${styles};
`;
export default StyledButton;

export { Button as ButtonVanilla };
