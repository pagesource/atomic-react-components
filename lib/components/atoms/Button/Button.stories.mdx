import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { primaryButton, primaryButtonDisabled, secondaryButton, secondaryButtonDisabled, customButtonLocator} from './Button.mock';
import Button, { ButtonVanilla } from '.';

<Meta title="Atoms/Buttons" component={Button} />

### Vanila  Button

<Preview>
  <Story name="Vanila Button">
   <ButtonVanilla {...primaryButton}>Submit</ButtonVanilla>
  </Story>
   <Story name="Vanila Disabled Button">
   <ButtonVanilla {...primaryButtonDisabled}>Submit</ButtonVanilla>
  </Story>
</Preview>


### Primary  Button

<Preview>
  <Story name="Primary Styled Button">
   <Button {...primaryButton}>Submit</Button>
  </Story>
   <Story name="Primary Styled Disabled Button">
   <Button {...primaryButtonDisabled}>Submit</Button>
  </Story>
</Preview>


### Secondary Button

<Preview>
  <Story name="Secondary Styled Button">
   <Button {...secondaryButton}>Submit</Button>
  </Story>
   <Story name="Secondary Styled Disabled Button">
   <Button {...secondaryButtonDisabled}>Submit</Button>
  </Story>
</Preview>

### custom Button Locator

<Preview>
  <Story name="Custom  Button">
   <Button {...customButtonLocator}>Submit</Button>
  </Story>
</Preview>

## Props ðŸ”§

<Props of={ButtonVanilla} />
