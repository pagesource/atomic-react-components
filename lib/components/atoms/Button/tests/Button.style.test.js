import React from 'react';
import { shallow } from 'enzyme';

import Button from '../Button';

import Theme from '../../../../styles/themes/base';

describe('Button', () => {
  let button = '';
  const theme = { ...Theme };

  test('should render correctly', () => {
    button = shallow(<Button theme={theme} />);
    expect(button).toMatchSnapshot();
  });
});
