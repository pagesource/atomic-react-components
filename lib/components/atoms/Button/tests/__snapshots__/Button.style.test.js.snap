// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
<StyledComponent
  forwardedComponent={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "Button-sc-1ccn477-0",
        "isStatic": false,
        "rules": Array [
          "border:none;border-radius:0;position:relative;min-height:48px;cursor:",
          [Function],
          ";",
          [Function],
          ";",
          [Function],
          ";",
          [Function],
          ";",
          [Function],
          ";&:focus:not(.focus-visible){outline:none;}&.icon{border:none;background-color:transparent;text-indent:-1000px;overflow:hidden;background-repeat:no-repeat;background-position:center;background-size:cover;min-width:20px;min-height:20px;padding:0;&.icon-image{background-image:url('",
          [Function],
          "');}}",
          [Function],
          ";",
          ";",
        ],
      },
      "displayName": "Button",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "Button-sc-1ccn477-0",
      "target": [Function],
      "toString": [Function],
      "usesTheme": true,
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    }
  }
  forwardedRef={null}
  theme={
    Object {
      "animations": Object {
        "ANIMATION_SLIDE_UP": "var(--animation-slide-up)",
        "ANIMATION_SLIDE_UP_WITH_MARGIN": "var(--animation-slide-up-with-margin)",
        "ANIMATION_SPINNER": "var(--animation-spinner)",
        "DURATION_M": "var(--duration-m)",
        "EASING_SWIFT": "var(--easing-swift)",
      },
      "behaviour": Object {
        "LINK_DECORATION": "text-decoration: none;",
      },
      "border": Object {
        "CTA_OUTLINE": "outline: var(--border-size-cta-tertiary) dashed var(--outline);",
        "CTA_TAB": "border-bottom: var(--border-size-cta-tertiary) solid var(--stroke);",
        "CTA_TAB_ACTIVE": "border-bottom: var(--border-size-cta-tertiary) solid var(--stroke-active);",
        "CTA_TAB_PANEL": "border-top: var(--border-size-cta-tertiary) solid var(--stroke);",
        "CTA_TAB_PANEL_VERTICAL": "border-left: var(--border-size-cta-tertiary) solid var(--stroke);",
        "CTA_TERTIARY": "
    border: var(--border-size-cta-tertiary) solid var(--border-secondary);
  ",
        "FIELD_ACTIVE": "
    border: var(--border-size-field-active) solid var(--field-border-active);
  ",
        "FIELD_DISABLED": "
    border: var(--border-size-field-active) solid var(--border-disabled);
  ",
        "FIELD_PRIMARY": "
    border: var(--border-size-field) solid var(--field-border-primary);
  ",
        "FIELD_RADIO": "
    border-radius: var(--border-radius-round);
  ",
        "FIELD_SECONDARY": "
    border: var(--border-size-field) solid var(--field-border-secondary);
  ",
        "LIST_INLINE": "
    border-right: var(--border-size-field) solid var(--field-border-primary);
  ",
      },
      "colors": Object {
        "CTA_DISABLED": "
    &[disabled]{
      background-color: var(--brand-disabled);
      color: var(--text-cta-disabled);
    }
  ",
        "CTA_PRIMARY": "
    background-color: var(--bg-cta-primary);
    color: var(--text-cta-primary);
    &:hover, &:active{
      background-color: var(--bg-cta-primary-active);
    }
  ",
        "CTA_SECONDARY": "
    background-color: var(--bg-cta-secondary);
    color: var(--text-cta-secondary);
    &:hover, &:active{
      background-color: var(--bg-cta-secondary-active);
    }
  ",
        "CTA_TAB": "
    background-color: var(--bg-cta-tertiary);
    color: var(--text-cta-tertiary);
  ",
        "CTA_TERTIARY": "
    background-color: var(--bg-cta-tertiary);
    color: var(--text-cta-tertiary);
    &:hover, &:active{
      background-color: var(--bg-cta-tertiary-active);
      color: var(--text-inverse);
    }
  ",
        "FIELD_DEFAULT": "
    color: var(--text-cta-tertiary);
    background-color: var(--bg-cta-tertiary);
  ",
        "FIELD_DISABLED": "
    background-color: var(--brand-disabled);
    color: 1px solid var(--text-cta-disabled);
  ",
        "LINK_PRIMARY": "
    color: var(--text-link);
  ",
        "TEXT_DEFAULT": "
    color: var(--text-primary);
  ",
        "TEXT_ERROR": "
    color: var(--error);
  ",
        "TEXT_INVERSE": "
    color: var(--text-inverse);
  ",
        "TEXT_LIGHT": "
    color: var(--text-light);
  ",
        "TEXT_SECONDARY": "
    color: var(--text-secondary);
  ",
        "TOGGLE_SWITCH": "
    background-color: var(--bg-cta-secondary);
  ",
      },
      "dimensions": Object {
        "CHECKBOX": "
    height: var(--spacing-3);
    width: var(--spacing-3);
  ",
        "CONTENT_MAX_WIDTH": "
    max-width: var(--spacing-max-content);
  ",
        "CTA": "
    height: var(--spacing-6);
    width: auto;
  ",
        "FIELD": "
    height: var(--spacing-6);
    width:100;
  ",
        "RADIO": "
    height: var(--spacing-3);
    width: var(--spacing-3);
  ",
        "TOGGLE_SWITCH": "
    height: var(--spacing-3);
    width: var(--spacing-6);
  ",
      },
      "icons": Object {
        "leftArrow": null,
        "pauseIcon": null,
        "playIcon": null,
        "rightArrow": null,
      },
      "shadow": Object {
        "SHADOW_L": "var(--shadow-l)",
        "SHADOW_M": "var(--shadow-m)",
        "SHADOW_S": "var(--shadow-s)",
      },
      "spacing": Object {
        "HALF_UNIT": "0.25rem",
        "MAX_CONTENT": "80rem",
        "QUAD_UNIT": "0.125rem",
        "UNIT_1": "0.5rem",
        "UNIT_10": "5rem",
        "UNIT_11": "5.5rem",
        "UNIT_12": "6rem",
        "UNIT_13": "6.5rem",
        "UNIT_14": "7rem",
        "UNIT_15": "7.5rem",
        "UNIT_16": "8rem",
        "UNIT_17": "8.5rem",
        "UNIT_18": "9rem",
        "UNIT_19": "9.5rem",
        "UNIT_2": "1rem",
        "UNIT_20": "10rem",
        "UNIT_3": "1.5rem",
        "UNIT_4": "2rem",
        "UNIT_5": "2.5rem",
        "UNIT_6": "3rem",
        "UNIT_7": "3.5rem",
        "UNIT_8": "4rem",
        "UNIT_9": "4.5rem",
      },
      "typography": Object {
        "FONT_CTA": "
    font-size: var(--font-size-cta);
    line-height: var(--line-height-cta);
  ",
        "FONT_DEFAULT": "
    font-family: var(--font-family);
    font-display: var(--font-family-display);
    font-size: var(--font-size-default);
    line-height: var(--line-height-default);
  ",
        "FONT_FIELD": "
    font-size: var(--font-size-field);
    line-height: var(--line-height-field);
  ",
        "FONT_H1": "
    font-family: var(--font-family-secondary);
    font-display: var(--font-family-display);
    font-size: var(--font-size-h1);
    line-height: var(--line-height-h1);
  ",
        "FONT_H2": "
    font-size: var(--font-size-h2);
    line-height: var(--line-height-h2);
  ",
        "FONT_H3": "
    font-size: var(--font-size-h3);
    line-height: var(--line-height-h3);
  ",
        "FONT_H4": "
    font-size: var(--font-size-h4);
    line-height: var(--line-height-h4);
  ",
        "FONT_LABEL": "
    font-size: var(--font-size-label);
    line-height: var(--line-height-label);
  ",
        "FONT_LINK": "
    font-size: var(--font-size-link);
    line-height: var(--line-height-link);
    text-decoration: none;
  ",
        "FONT_QUOTE_SIGN": "
    font-size: var(--font-size-quote);
    line-height: var(--line-height-quote);
  ",
      },
      "zIndex": Object {
        "zFlyOut": "z-index: var(--z-index-fly-out)",
        "zLoader": "z-index: var(--z-index-loader)",
        "zModal": "z-index: var(--z-index-modal)",
        "zNativeTickSelectIcon": "z-index: var(--z-index-native-tick-select-icon)",
        "zOverlay": "z-index: var(--z-index-overlay)",
        "zSticky": "z-index: var(--z-index-sticky)",
      },
    }
  }
/>
`;
